tardis/utilities/asyncbulkcall.py

```diff
@@ -0,0 +1,173 @@
+                f", got {self._concurrency!r} instead"
+            )
+
+    async def __call__(self, __task: T) -> R:
+        """Execute a ``task`` in bulk and return the result"""
+        result: "asyncio.Future[R]" = asyncio.get_event_loop().create_future()
+        await self._queue.put((__task, result))
+        self._ensure_worker()
+        return await result
+
+    def _ensure_worker(self):
+        """Ensure there is a worker to dispatch tasks for command execution"""
+        if self._master_worker is None:
+            self._master_worker = asyncio.ensure_future(self._bulk_dispatch())
+
+    async def _bulk_dispatch(self):
+        """Collect tasks into bulks and dispatch them for command execution"""
+        while True:
+            await asyncio.sleep(0)
+            bulk = list(zip(*(await self._get_bulk())))
+            if not bulk:
+                continue
+            tasks, futures = bulk
+            await self._concurrent.acquire()
+            asyncio.ensure_future(self._bulk_execute(tuple(tasks), futures))
+
+    async def _get_bulk(self) -> "List[Tuple[T, asyncio.Future[R]]]":
+        """Fetch the next bulk from the internal queue"""
+        results = []
+        max_items, queue = self._size, self._queue
+        deadline = time.monotonic() + self._delay
+        while len(results) < max_items:
+            try:
+                if queue.empty():
+                    item = await asyncio.wait_for(
+                        queue.get(), deadline - time.monotonic()
+                    )
+                else:
+                    item = queue.get_nowait()
+            except asyncio.TimeoutError:
+                break
+            else:
+                results.append(item)
+                queue.task_done()
+            # check deadline late so that we cannot stall if the delay is very low
+            if time.monotonic() > deadline:
+                break
+        return results
+
+    async def _bulk_execute(
+        self, tasks: Tuple[T, ...], futures: "List[asyncio.Future[R]]"
+    ) -> None:
+        """Execute several ``tasks`` in bulk and set their ``futures``' result"""
+        try:
+            results = await self._command(*tasks)
+            # make sure we can cleanly match input to output
+            results = [None] * len(futures) if results is None else list(results)
+            if len(results) != len(futures):
+                raise RuntimeError(
+                    f"bulk command {self._command} provided {len(results)} results"
+                    f", expected {len(futures)} results or 'None'"
+                )
+        except Exception as task_exception:
+            for future in futures:
+                future.set_exception(task_exception)
+        else:
+            for future, result in zip(futures, results):
+                future.set_result(result)
+        finally:
+            self._concurrent.release()
```

Comment:
It seems a bit out of context to release `self._concurrent` here as it was created somewhere else. This hardens extending/maintaining the code as one first needs to check where and how it is handled. It would be great if you could check to improve the workflow of handling this :)

Reply:
I was able to push this up into ``_bulk_dispatch``. It's still a bit less straightforward than I would like but at least it is all in one place.